scripts/run_action.mjs

```diff
@@ -14,87 +14,169 @@
  * It runs an `*.action.mjs` or `*.action.sh` file with the given parameters.
  * You can find these files located within their associated directories in the project.
- * Action files can also define:
- *   `requirements` - actions that must be run beforehand as prerequisite.
- *   `dependencies` - files and folders that the action takes as input. If they remain unchanged, the action can be skipped.
  *
  * @param {string} actionPath The truncated path to the action you wish to run (e.g. "www/build")
- * @param {...string} parameters The flags and other parameters we want to run the action with.
+ * @param {Object} options Additional action options
+ * @param {string[]} options.parameters Command line arguments passed in to the action script
+ * @param {string[]} options.inputs Files and folders the action consumes. Specify these to skip the action when able
  * @returns {void}
  */
-export async function runAction(actionPath, ...parameters) {
-  const resolvedPath = await resolveActionPath(actionPath);
+export async function runAction(actionPath, {parameters = [], inputs = []} = {}) {
+  /**
+   * @description promisifies the child process spawner
+   */
+  const spawnPromise = (command, parameters) =>
+    new Promise((resolve, reject) => {
+      const childProcess = spawn(command, parameters, {shell: true});
+
+      const forEachMessageLine = (buffer, callback) => {
+        buffer
+          .toString()
+          .split('\n')
+          .filter(line => line.trim())
+          .forEach(callback);
+      };
+
+      childProcess.stdout.on('data', data => forEachMessageLine(data, line => console.info(line)));
+      childProcess.stderr.on('data', error => forEachMessageLine(error, line => console.error(chalk.red(line))));
+
+      childProcess.on('close', code => {
+        if (code === 0) {
+          resolve(childProcess);
+        } else {
+          reject(childProcess);
+        }
+      });
+    });
+
+  let resolvedPath;
+  const pathRoots = [process.env.ROOT_DIR, path.join(process.env.ROOT_DIR, 'src')];
+  const fileExtensions = ['sh', 'mjs'];
+
+  for (const root of pathRoots) {
+    for (const extension of fileExtensions) {
+      const pathCandidate = `${path.resolve(root, actionPath)}.action.${extension}`;
+
+      if (existsSync(pathCandidate)) {
+        resolvedPath = pathCandidate;
+        break;
+      }
+    }
+  }
+
   if (!resolvedPath) {
     console.info('Please provide an action to run.');
     return runAction('list');
   }
 
+  const cache = new ActionCache();
+
+  const actionOptions = {parameters, inputs};
+  const lastRunMetadata = await cache.read(resolvedPath, actionOptions);
+
+  if (inputs.length) {
+    for (const input of inputs) {
+      if (!existsSync(input)) {
+        throw new ReferenceError(`action(${actionPath}) requires the input: "${input}". Aborting!`);
+      }
+    }
+
+    let mostRecentlyModifiedInput = 0;
+    let inputsToCheck = [...inputs];
```

Comment:
This only considers direct inputs.

If you have f1 -> action1 -> f2 -> action2 -> f3

If f1 changes and you run action2, you will miss the update to f1.

You need to navigate the entire tree of dependencies. You need to check not only the direct dependencies, but also the dependencies of the actions you call. It's a two-pass approach.

I guess the logic is:
1. (Always) Run all your direct dependencies
2. If any direct dependency got updated, run the action

Build systems are not easy. I wonder if it makes sense to use NX or bazel instead, or even make.
Perhaps it's worth playing with what they would look like at some point, before investing too much on our own build system.

Reply:
I don't understand how this doesn't do that! In this PR, each action when called is checking its own dependencies. So we effectively navigate the entire tree of dependencies manually and recursively! For instance:

https://github.com/Jigsaw-Code/outline-client/pull/1291/files#diff-3ade035b399e6b2f6083fc0cbee893d3e5683f71af90d77c5e7a70837ca5bd23R31-R32

The first `runAction` call on line 31 may create a new `www`. This `www` is explicitly passed into the input call in the second action call of `runAction` on line 32. So that call will see if that folder changes and be called anyway. And this all is tracked within the top level action we're in - it makes it very clear (to me, perhaps it needs some tweaking) what depends on which when I read the action file and what the chain of dependencies are.

Reply:
As discussed in our meeting, it's important to make sure the build is correct.
Even if it's correct with the existing actions, changes and new actions can break without warning. That has caused us significant pain in the past.

I believe this can be fixed by checking the transitive inputs first.
Alternatively, instead of having an actions cache in the build system, specific actions could be made incremental. Similar to `cordova prepare`, but taking into account the inputs properly.

Reply:
Actually, thinking about it over the weekend - what if we stopped thinking about this as a build system and more as a "scripting framework"?

The reason this currently works, and why the `'cordova/setup'` call here -

```js
  await runAction('www/build', {parameters: [`--buildMode=${buildMode}`], inputs: ['src/www']});
  await runAction('cordova/setup', {parameters, inputs: ['www', 'resources', 'src/cordova', 'cordova-plugin-outline']});
```

- doesn't have to also include `"src/www"` as an input is because [`cordova/setup` itself calls the `"www/build"` action as well with `"src/www"` as input](https://github.com/Jigsaw-Code/outline-client/pull/1291/files#diff-83e4918643471872d89a312fdd8603e164f6bc00c21c6bce3c877b07821898e0R52). Basically the way I have it set up now, if each action calls _all_ its child actions directly, it works. Yes, this results in some redundant code, but it's no better than what we had before. In fact we could have each action declare "input defaults" and then we wouldn't even need to think about what inputs are needed anymore. I can even add an option `skipIfNoInputChanges` to make it opt-in and more clear what's happening when you read the file.

I actually like this "each action calls all its child actions" approach because it allows us to specify when we write a new action how its dependencies should be called. Imagine an action that builds all platforms - if we statically declare a list of actions we need to call, we lose flexibility of potentially being able to build and rmraf `www` between platform builds. I don't think Nx could give us something like that.

That said, I'm not against refactoring this system just that each action DFSes its child components to determine skipability, however - turning actions into extensible classes would require typescript interfaces, which would require moving everything to `ts-node` - and at that point I'm thinking maybe the work is similar in scope to trying to add a bunch of `package.json`s to our `src` directory and leveraging workspace tools?

Basically, is "each action must explicitly call all its child actions" sufficient enough of a rule that we can get the time savings of this PR for now and refactor things into our ideal state a bit later?